### PR TITLE
Rutusamai/Show Build Task properties

### LINKED ISSUE #61: Enable Show Task Configuration
With a build-task selected, I should be able to right-click and show/edit the task definition.

### PR DESCRIPTION
**Pull request by rsamai**
-Click on Build Tasks under a registry, Right Click on a Build Task and select "Show Build Task Properties"
-Can Invoke from Command Palette also with new quick pick build task function
-JSON file opens in a new tab with properties showing
-Doesn't use the Azure CLI 
-Closes #61

### CODE REVIEW

File: commands/utils/quick-pick-azure.ts [hunk 1]
```diff
@@ -33,6 +34,23 @@ export async function quickPickACRRepository(registry: Registry, prompt?: string
...
+    const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
+
+    let buildTasks: ContainerModels.BuildTask[] = [];
+    let buildTaskNames: string[] = [];
+    buildTasks = await client.buildTasks.list(resourceGroup.name, registry.name);
+
+    for (let bt of buildTasks) {
+        buildTaskNames.push(bt.name);
+    }
+    let desiredBuildTask = await vscode.window.showQuickPick(buildTaskNames, { 'canPickMany': false, 'placeHolder': placeHolder });
```
Comment (estebanreyl) [inline commands/utils/quick-pick-azure.ts]: use ext.ui.showQuickPick, see the others to have a basis on how to do it. This does consistent error handling and has the added benefit of having the recently selected thingy if you wrap items in the azurequickpick item type.

File: commands/utils/quick-pick-azure.ts [hunk 2]
```diff
@@ -33,6 +34,23 @@ export async function quickPickACRRepository(registry: Registry, prompt?: string
...
+    if (!desiredBuildTask) { return; }
+    const buildTask = buildTasks.find((currentBuildTask): boolean => { return desiredBuildTask === currentBuildTask.name });
```
Comment (estebanreyl) [inline commands/utils/quick-pick-azure.ts]: just return the find? you can avoid this by following the abovementioned model.

File: commands/azureCommands/show-buildTask.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
+    let item: any = await client.buildTasks.get(resourceGroup.name, registry.name, buildTask);
+
+    try {
+        let steps = await client.buildSteps.get(resourceGroup.name, registry.name, buildTask, `${buildTask}StepName`);
+        item.properties = steps;
+    } catch (error) {
+        console.error("Build Step not available for this image due to update in API");
```
Comment (estebanreyl) [inline commands/azureCommands/show-buildTask.ts]: console.error the actual caught error ? That may give more specific information

File: commands/azureCommands/show-buildTask.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        registry = await quickPickACRRegistry();
+        resourceGroup = await acrTools.getResourceGroup(registry, subscription);
+        buildTask = (await quickPickBuildTask(registry, subscription, resourceGroup)).name;
+    }
+
+    const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
+    let item: any = await client.buildTasks.get(resourceGroup.name, registry.name, buildTask);
+
+    try {
+        let steps = await client.buildSteps.get(resourceGroup.name, registry.name, buildTask, `${buildTask}StepName`);
```
Comment (estebanreyl) [inline commands/azureCommands/show-buildTask.ts]: const seems more appropiate

File: commands/azureCommands/show-buildTask.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    let subscription: Subscription;
+    let registry: Registry;
+    let resourceGroup: ResourceGroup;
+    let buildTask: string;
+
+    if (context) { // Right click
+        subscription = context.susbscription;
+        registry = context.registry;
+        resourceGroup = await acrTools.getResourceGroup(registry, subscription);
+        buildTask = context.label;
```
Comment (estebanreyl) [inline commands/azureCommands/show-buildTask.ts]: Don't use context.label, we could potentially show the tasks differently at some point, instead store the acquired task information in the nodes.

File: package.json [hunk 6]
```diff
@@ -610,6 +627,11 @@
         "title": "Browse in the Azure Portal",
         "category": "Docker"
       },
+      {
+        "command": "vscode-docker.showBuildTaskProperties",
```
Comment (estebanreyl) [inline package.json]: Remember to use standardized names similar to the create registry, etc.

File: commands/azureCommands/show-buildTask.ts (hunk 3 above)
Comment (rsamai) [reply commands/azureCommands/show-buildTask.ts]: Added both

File: explorer/models/taskNode.ts [hunk 7]
```diff
@@ -1,21 +1,23 @@
...
+    private _onDidChangeTreeData: vscode.EventEmitter<NodeBase> = new vscode.EventEmitter<NodeBase>();
+    public readonly onDidChangeTreeData: vscode.Event<NodeBase> = this._onDidChangeTreeData.event;
     constructor(
         public readonly label: string,
-        public readonly contextValue: string,
         public subscription: SubscriptionModels.Subscription,
         public readonly azureAccount: AzureAccount,
         public registry: ContainerModels.Registry,
-        public readonly iconPath: any = {}
+        public readonly iconPath: any = {},
```
Comment (northtyphoon) [inline explorer/models/taskNode.ts]: The code in getTreeItem uses null for iconPath. I guess both `{}` or `null` work. Can you update them to be consistent?

File: commands/azureCommands/task-utils/showTaskManager.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        return this.onDidChangeEvent.event;
+    }
+
+    public update(uri: vscode.Uri, message: string): void {
+        this.onDidChangeEvent.fire(uri);
+    }
+}
+
+export function decodeBase64(str: string): string {
+    return Buffer.from(str, 'base64').toString('ascii');
```
Comment (northtyphoon) [inline commands/azureCommands/task-utils/showTaskManager.ts]: Does it support unicode? If yes, should you consider `toString('utf8')`?

File: commands/azureCommands/show-buildTask.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    try {
+        const steps = await client.buildSteps.get(resourceGroup.name, registry.name, buildTask, `${buildTask}StepName`);
+        item.properties = steps;
+    } catch (error) {
+        console.error(error);
+        console.error("Build Step not available for this image due to update in API");
+    }
+
+    openTask(JSON.stringify(item, undefined, 1), buildTask);
```
Comment (northtyphoon) [inline commands/azureCommands/show-buildTask.ts]: Can you use named parameters for `undefined` and `1` for readability?

File: commands/azureCommands/show-buildTask.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
+    let item: any = await client.buildTasks.get(resourceGroup.name, registry.name, buildTask);
+
+    try {
+        const steps = await client.buildSteps.get(resourceGroup.name, registry.name, buildTask, `${buildTask}StepName`);
+        item.properties = steps;
+    } catch (error) {
+        console.error(error);
```
Comment (northtyphoon) [inline commands/azureCommands/show-buildTask.ts]: Do we still use console.error in current code base for debug?

File: commands/azureCommands/show-buildTask.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        registry = await quickPickACRRegistry();
+        resourceGroup = await acrTools.getResourceGroup(registry, subscription);
+        buildTask = (await quickPickBuildTask(registry, subscription, resourceGroup)).name;
+    }
+
+    const client = AzureUtilityManager.getInstance().getContainerRegistryManagementClient(subscription);
+    let item: any = await client.buildTasks.get(resourceGroup.name, registry.name, buildTask);
+
+    try {
+        const steps = await client.buildSteps.get(resourceGroup.name, registry.name, buildTask, `${buildTask}StepName`);
```
Comment (northtyphoon) [inline commands/azureCommands/show-buildTask.ts]: We probably should not assume the build step name is always `${buildTask}StepName`. Does CLI code also hard code the step name ? We will get ride of build step in GA API. So I'm fine to leave with it.

Comment (rsamai) [reply commands/azureCommands/show-buildTask.ts]: Yeah, that's how the CLI does it

File: commands/azureCommands/show-buildTask.ts (hunk 10 above)
Comment (rsamai) [reply commands/azureCommands/show-buildTask.ts]: Changed to ext.outputChannel.append(error);

File: commands/azureCommands/task-utils/showTaskManager.ts (hunk 8 above)
Comment (rsamai) [reply commands/azureCommands/task-utils/showTaskManager.ts]: Changed it to that

**Pull request merged**